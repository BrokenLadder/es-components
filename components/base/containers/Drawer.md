Use ``Drawer`` and ``DrawerPanel`` components to wrap content in collapsable panels.

### Basic Example:
```
<Drawer>
  <DrawerPanel header="Collapsable Item 1">
    <p>Pretty much any content you want in here.</p>
  </DrawerPanel>
  <DrawerPanel header="Collapsable Item 2">
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </DrawerPanel>
  <DrawerPanel header="Collapsable Item 3">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </DrawerPanel>
</Drawer>
```

### Drawer Options

Option            | Type                   | Description
----------------- | ---------------------- | ---------------------------------------------------------
isAccordion       | boolean                | Only allows one DrawerPanel to be open at a time
defaultActiveKeys | string or string array | Specify which panels are opened by default
closedIconName    | string                 | Override the default plus icon with another OE icon name
openedIconName    | string                 | Override the default minus icon with another OE icon name
className         | string or object       | Add additional CSS classes to the root drawer element

### DrawerPanel Options

Option    | Type             | Description
--------- | ---------------- | ---------------------------------------------------------------
header    | string           | Header text displayed next to the open/close icon
key       | string           | Manually set the key value for the panel (otherwise uses index)
className | string or object | Add additional CSS classes to the drawer item element
noPadding | boolean          | Removes the default padding from the panel body


### Advanced Example:
```
<Drawer isAccordion defaultActiveKeys="1" closedIconName="hand-right" openedIconName="hand-down" className="drawer-big">
  <DrawerPanel header="Collapsable Item 1" key="1">
    <p>Pretty much any content you want in here.</p>
  </DrawerPanel>
  <DrawerPanel header="Collapsable Item 2" key="2" className="alt-color">
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </DrawerPanel>
  <DrawerPanel header="Collapsable Item 3" key="3" noPadding>
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </DrawerPanel>
</Drawer>
```
