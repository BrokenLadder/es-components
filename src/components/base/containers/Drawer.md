Use ``Drawer`` and ``Drawer.Panel`` components to wrap content in collapsable panels.

### Basic Example:
```
<Drawer>
  <Drawer.Panel title="Panel 1">
    <p>Pretty much any content you want in here.</p>
  </Drawer.Panel>
  <Drawer.Panel title="Panel 2">
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </Drawer.Panel>
  <Drawer.Panel title="Panel 3">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </Drawer.Panel>
</Drawer>
```

### Advanced Example:

Use the ``noPadding`` property to remove default padding within a panel (useful for lists and tables).
You can customize the key property of each Panel; if not specified the default is a zero-based array.
Both the Drawer and Drawer.Panel components will accept additional classNames.

```
<Drawer isAccordion defaultActiveKeys="1" closedIconName="hand-right" openedIconName="hand-down" className="drawer-big">
  <Drawer.Panel title="Panel 1" key="1">
    <p>Pretty much any content you want in here.</p>
  </Drawer.Panel>
  <Drawer.Panel title="Panel 2" key="2" noPadding>
    <ul>
      <li>List Item</li>
      <li>Another Item</li>
    </ul>
  </Drawer.Panel>
  <Drawer.Panel title="Panel 3" key="3" className="alt-color">
    <div>
      More content. Anim pariatur cliche reprehenderit, enim eiusmod
      high life accusamus terry richardson ad squid.
    </div>
  </Drawer.Panel>
</Drawer>
```
