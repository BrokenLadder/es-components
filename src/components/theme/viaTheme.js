const white = '#fff';
const gray1 = '#f4f4f4';
const gray2 = '#ecedee';
const gray3 = '#e4e4e4';
const gray4 = '#d8d8d8';
const gray5 = '#c5c5c5';
const gray6 = '#979797';
// const gray7 = '#707070';
const gray8 = '#585858';
const gray9 = '#444';
const black = '#000';

const grayLightest = gray1;
const grayLighter = gray2;
const grayLight = gray3;
const gray = gray5;
const grayDark = gray6;
const grayDarker = gray8;
const grayDarkest = gray9;

const dflt = '#d8d8d8';
const dfltHover = '#bfbfbf';
const primary = '#007fa7';
const primaryHover = '#005874';
const info = '#069';
const infoHover = '#004466';
const success = '#298544';
const successHover = '#1d5e30';
const warning = '#b35f00';
const warningHover = '#804400';
const danger = '#c00';
const dangerHover = '#990000';
const advisor = warning;

const boxShadowLight = 'rgba(0, 0, 0, 0.075)';
const boxShadowDark = 'rgba(0, 0, 0, 0.5)';

const inputBorder = gray4;
const inputFocus = '#83bffc';
const inputBoxShadow = 'rgba(102, 175, 233, 0.6)';

// Datepicker colors
const navArrow = gray;
const navArrowHover = grayDark;
const selected = '#216ba5';
const hover = '#1d5d90';
const keyboard = '#2a87d0';
const inRange = 'rgba(33, 107, 165, 0.5)';
const highlight = '#3dcc4a';
const highlightHover = '#32be3f';

const theme = {
  colors: {
    black,
    grayDarkest,
    grayDarker,
    grayDark,
    gray,
    grayLight,
    grayLighter,
    grayLightest,
    white,
    dflt,
    dfltHover,
    primary,
    primaryHover,
    success,
    successHover,
    info,
    infoHover,
    warning,
    warningHover,
    danger,
    dangerHover,
    advisor,
    boxShadowLight,
    boxShadowDark,
    inputFocus
  },
  datepickerColors: {
    navArrow,
    navArrowHover,
    selected,
    hover,
    keyboard,
    inRange,
    highlight,
    highlightHover
  },
  buttonStyles: {
    buttonsNormal: {
      primary: {
        bgColor: primary,
        textColor: white,
        hoverBgColor: primaryHover,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        boxShadowColor: primaryHover
      },
      secondary: {
        bgColor: dflt,
        textColor: white,
        hoverBgColor: dfltHover,
        hoverTextColor: white,
        activeBgColor: dfltHover,
        activeTextColor: white,
        boxShadowColor: dfltHover
      },
      success: {
        bgColor: success,
        textColor: white,
        hoverBgColor: successHover,
        hoverTextColor: white,
        activeBgColor: successHover,
        activeTextColor: white,
        boxShadowColor: successHover
      },
      info: {
        bgColor: info,
        textColor: white,
        hoverBgColor: infoHover,
        hoverTextColor: white,
        activeBgColor: infoHover,
        activeTextColor: white,
        boxShadowColor: infoHover
      },
      warning: {
        bgColor: warning,
        textColor: white,
        hoverBgColor: warningHover,
        hoverTextColor: white,
        activeBgColor: warningHover,
        activeTextColor: white,
        boxShadowColor: warningHover
      },
      danger: {
        bgColor: danger,
        textColor: white,
        hoverBgColor: dangerHover,
        hoverTextColor: white,
        activeBgColor: dangerHover,
        activeTextColor: white,
        boxShadowColor: dangerHover
      }
    },
    buttonsAlt: {
      primary: {
        bgColor: white,
        textColor: primary,
        hoverBgColor: primary,
        hoverTextColor: white,
        activeBgColor: primaryHover,
        activeTextColor: white,
        borderColor: primary
      },
      gray: {
        bgColor: white,
        textColor: '#63666a',
        hoverBgColor: '#63666a',
        hoverTextColor: white,
        activeBgColor: '#4a4d50',
        activeTextColor: white,
        borderColor: '#63666a'
      },
      magenta: {
        bgColor: white,
        textColor: '#c110a0',
        hoverBgColor: '#c110a0',
        hoverTextColor: white,
        activeBgColor: '#920C79',
        activeTextColor: white,
        borderColor: '#c110a0'
      },
      violet: {
        bgColor: white,
        textColor: '#702082',
        hoverBgColor: '#702082',
        hoverTextColor: white,
        activeBgColor: '#4d1659',
        activeTextColor: white,
        borderColor: '#702082'
      }
    }
  },
  // buttonSizes should always have default, lg, sm, xs
  buttonSizes: {
    lg: {
      borderRadius: '3px',
      fontSize: '26px',
      padding: '8px 20px 5px'
    },
    default: {
      borderRadius: '2px',
      fontSize: '18px',
      padding: '5px 15px 4px'
    },
    sm: {
      borderRadius: '2px',
      fontSize: '14px',
      padding: '4px 10px 3px'
    },
    xs: {
      borderRadius: '2px',
      fontSize: '12px',
      padding: '3px 10px 2px'
    }
  },
  validationIconName: {
    success: 'ok',
    warning: 'warning-sign',
    danger: 'remove'
  },
  validationTextColor: {
    success,
    warning: warningHover,
    danger
  },
  validationInputColor: {
    success: {
      borderColor: success,
      boxShadow: `inset 0 0 5px ${success}`,
      focusBorderColor: success,
      focusBoxShadow: `0 0 6px ${success}`
    },
    warning: {
      borderColor: warning,
      boxShadow: `inset 0 0 5px ${warning}`,
      focusBorderColor: warning,
      focusBoxShadow: `0 0 6px ${warning}`
    },
    danger: {
      borderColor: danger,
      boxShadow: `inset 0 0 5px ${danger}`,
      focusBorderColor: danger,
      focusBoxShadow: `0 0 6px ${danger}`
    },
    default: {
      borderColor: inputBorder,
      boxShadow: `inset 0 1px 1px ${boxShadowLight}`,
      focusBorderColor: inputFocus,
      focusBoxShadow: `0 0 8px ${inputBoxShadow}`
    }
  },
  sizes: {
    baseFontSize: 18,
    baseLineHeight: 24
  },
  screenSize: {
    retina: '320px',
    phone: '480px',
    tablet: '768px',
    desktop: '992px',
    widescreen: '1200px'
  }
};

export default theme;
