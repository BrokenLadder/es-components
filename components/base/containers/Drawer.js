import React, { PropTypes, Children, Component } from 'react';
import classNames from 'classnames';
import './drawer.less';

function toArray(activeKey) {
  let currentActiveKey = activeKey;
  if (!Array.isArray(currentActiveKey)) {
    currentActiveKey = currentActiveKey ? [currentActiveKey] : [];
  }
  return currentActiveKey;
}

class Drawer extends Component {
  constructor(props) {
    super(props);

    const currentActiveKey = this.props.defaultActiveKeys;
    this.state = {
      activeKey: toArray(currentActiveKey)
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('activeKey' in nextProps) {
      this.setState({
        activeKey: toArray(nextProps.activeKey)
      });
    }
  }

  onClickItem(key) {
    return () => {
      let activeKey = this.state.activeKey;

      if (this.props.isAccordion) {
        activeKey = activeKey[0] === key ? [] : [key];
      } else {
        activeKey = [...activeKey];
        const index = activeKey.indexOf(key);
        const isActive = index > -1;
        if (isActive) {
          activeKey.splice(index, 1);
        } else {
          activeKey.push(key);
        }
      }

      this.setState({ activeKey });
    };
  }

  getItems() {
    const activeKey = this.state.activeKey;
    const { isAccordion } = this.props;
    const newChildren = [];

    Children.forEach(this.props.children, (child, index) => {
      if (!child) return;
      // If there is no key provided, use the panel order as default key
      const key = child.key || String(index);
      const header = child.props.header;

      let isActive = false;
      if (isAccordion) {
        isActive = activeKey[0] === key;
      } else {
        isActive = activeKey.indexOf(key) > -1;
      }

      const props = {
        key,
        header,
        isActive,
        children: child.props.children,
        onItemClick: this.onClickItem(key).bind(this),
        closedIconName: this.props.closedIconName,
        openedIconName: this.props.openedIconName
      };

      newChildren.push(React.cloneElement(child, props));
    });

    return newChildren;
  }

  render() {
    const { className } = this.props;
    const classes = classNames('drawer', className);

    return (
      <div className={classes}>
        {this.getItems()}
      </div>
    );
  }
}

Drawer.propTypes = {
  children: PropTypes.any,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object
  ]),
  closedIconName: PropTypes.string,
  defaultActiveKeys: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string)
  ]),
  isAccordion: PropTypes.bool,
  openedIconName: PropTypes.string
};

Drawer.defaultProps = {
  closedIconName: 'plus',
  isAccordion: false,
  openedIconName: 'minus'
};

export default Drawer;
