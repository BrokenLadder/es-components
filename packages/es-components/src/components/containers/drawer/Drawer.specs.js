/* eslint-env jest */

import React from 'react';
import { render, cleanup } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import viaTheme from 'es-components-via-theme';

import Drawer from './Drawer';

jest.mock('../../util/generateAlphaName', () => () => 'abcdef');

const buildDrawer = props => (
  <ThemeProvider theme={viaTheme}>
    <Drawer className="important" {...props}>
      <Drawer.Panel
        title="collapse 1"
        key="1"
        className="first"
        titleAside="side text"
      >
        first
      </Drawer.Panel>
      <Drawer.Panel title="collapse 2" key="2" className="second" noPadding>
        second
      </Drawer.Panel>
      <Drawer.Panel title="collapse 3" key="3" className="third">
        third
      </Drawer.Panel>
    </Drawer>
  </ThemeProvider>
);

beforeEach(cleanup);

describe('drawer', () => {
  it('renders as expected', () => {
    const { container } = render(buildDrawer());
    expect(container).toMatchSnapshot();
  });

  it('opens a closed panel', () => {
    const onActiveKeysChanged = jest.fn();
    const { getByText } = render(buildDrawer({ onActiveKeysChanged }));

    getByText('collapse 1').click();

    expect(getByText('first')).toBeVisible();
    expect(getByText('second')).not.toBeVisible();
    expect(getByText('third')).not.toBeVisible();
  });

  it('allows multiple panels to be opened at the same time', () => {
    const onActiveKeysChanged = jest.fn();
    const { getByText } = render(buildDrawer({ onActiveKeysChanged }));

    getByText('collapse 1').click();
    getByText('collapse 3').click();

    expect(getByText('first')).toBeVisible();
    expect(getByText('second')).not.toBeVisible();
    expect(getByText('third')).toBeVisible();
  });

  it('closes a previously opened panel', async () => {
    jest.useFakeTimers();
    const onActiveKeysChanged = jest.fn();
    const { getByText } = render(buildDrawer({ onActiveKeysChanged }));

    const secondPanelToggle = getByText('collapse 2');
    secondPanelToggle.click();

    expect(getByText('second')).toBeVisible();

    secondPanelToggle.click();
    jest.runOnlyPendingTimers();
    expect(getByText('second')).not.toBeVisible();
  });

  it('allows drawers to be opened by default', () => {
    const onActiveKeysChanged = jest.fn();
    const { getByText } = render(
      buildDrawer({ activeKeys: ['1', '3'], onActiveKeysChanged })
    );

    expect(getByText('first')).toBeVisible();
    expect(getByText('second')).not.toBeVisible();
    expect(getByText('third')).toBeVisible();
  });
});

describe('accordion', () => {
  const isAccordion = true;

  it('renders as expected', () => {
    const { container } = render(buildDrawer({ isAccordion }));
    expect(container).toMatchSnapshot();
  });

  it('should only allow one drawer to be opened at a time', () => {
    jest.useFakeTimers();
    const onActiveKeysChanged = jest.fn();
    const { getByText } = render(
      buildDrawer({ isAccordion, onActiveKeysChanged, activeKeys: ['1'] })
    );

    expect(getByText('first')).toBeVisible();

    getByText('collapse 2').click();
    jest.runOnlyPendingTimers();

    expect(getByText('second')).not.toBeVisible();
  });
});
