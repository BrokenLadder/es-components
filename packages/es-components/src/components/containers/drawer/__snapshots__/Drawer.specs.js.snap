// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawer component accordion renders as expected 1`] = `
<div
  className="es-drawer important sc-EHOje ixLrxb"
>
  <div
    className="es-drawer__panel first sc-bwzfXH bBIqkK"
  >
    <div
      className="es-drawer__heading"
      id="abcdef"
      role="heading"
    >
      <button
        aria-controls="abcdef"
        aria-expanded={false}
        className="sc-htpNat jDwJgO"
        onClick={[Function]}
      >
        <span>
          <i
            aria-hidden={true}
            className="es-icon sc-bxivhb bCiBaT sc-bdVaJa eTkFwb"
            content="601"
            fontFamily="indv-mkt-icons"
            fontSize="inherit"
          />
          collapse 1
        </span>
        <aside>
          side text
        </aside>
      </button>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="abcdef"
      className="rah-static rah-static--height-zero es-drawer__body sc-ifAKCX HqBXT"
      id="abcdef"
      role="region"
      style={
        Object {
          "height": 0,
          "overflow": "hidden",
        }
      }
    >
      <div>
        first
      </div>
    </div>
  </div>
  <div
    className="es-drawer__panel second sc-bwzfXH bBIqkK"
  >
    <div
      className="es-drawer__heading"
      id="abcdef"
      role="heading"
    >
      <button
        aria-controls="abcdef"
        aria-expanded={false}
        className="sc-htpNat jDwJgO"
        onClick={[Function]}
      >
        <span>
          <i
            aria-hidden={true}
            className="es-icon sc-bxivhb bCiBaT sc-bdVaJa eTkFwb"
            content="601"
            fontFamily="indv-mkt-icons"
            fontSize="inherit"
          />
          collapse 2
        </span>
        <aside />
      </button>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="abcdef"
      className="rah-static rah-static--height-zero es-drawer__body sc-ifAKCX bUSXbP"
      id="abcdef"
      role="region"
      style={
        Object {
          "height": 0,
          "overflow": "hidden",
        }
      }
    >
      <div>
        second
      </div>
    </div>
  </div>
  <div
    className="es-drawer__panel third sc-bwzfXH bBIqkK"
  >
    <div
      className="es-drawer__heading"
      id="abcdef"
      role="heading"
    >
      <button
        aria-controls="abcdef"
        aria-expanded={false}
        className="sc-htpNat jDwJgO"
        onClick={[Function]}
      >
        <span>
          <i
            aria-hidden={true}
            className="es-icon sc-bxivhb bCiBaT sc-bdVaJa eTkFwb"
            content="601"
            fontFamily="indv-mkt-icons"
            fontSize="inherit"
          />
          collapse 3
        </span>
        <aside />
      </button>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="abcdef"
      className="rah-static rah-static--height-zero es-drawer__body sc-ifAKCX HqBXT"
      id="abcdef"
      role="region"
      style={
        Object {
          "height": 0,
          "overflow": "hidden",
        }
      }
    >
      <div>
        third
      </div>
    </div>
  </div>
</div>
`;

exports[`drawer component drawer renders as expected 1`] = `
<div
  className="es-drawer important sc-EHOje ixLrxb"
>
  <div
    className="es-drawer__panel first sc-bwzfXH bBIqkK"
  >
    <div
      className="es-drawer__heading"
      id="abcdef"
      role="heading"
    >
      <button
        aria-controls="abcdef"
        aria-expanded={false}
        className="sc-htpNat jDwJgO"
        onClick={[Function]}
      >
        <span>
          <i
            aria-hidden={true}
            className="es-icon sc-bxivhb bCiBaT sc-bdVaJa eTkFwb"
            content="601"
            fontFamily="indv-mkt-icons"
            fontSize="inherit"
          />
          collapse 1
        </span>
        <aside>
          side text
        </aside>
      </button>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="abcdef"
      className="rah-static rah-static--height-zero es-drawer__body sc-ifAKCX HqBXT"
      id="abcdef"
      role="region"
      style={
        Object {
          "height": 0,
          "overflow": "hidden",
        }
      }
    >
      <div>
        first
      </div>
    </div>
  </div>
  <div
    className="es-drawer__panel second sc-bwzfXH bBIqkK"
  >
    <div
      className="es-drawer__heading"
      id="abcdef"
      role="heading"
    >
      <button
        aria-controls="abcdef"
        aria-expanded={false}
        className="sc-htpNat jDwJgO"
        onClick={[Function]}
      >
        <span>
          <i
            aria-hidden={true}
            className="es-icon sc-bxivhb bCiBaT sc-bdVaJa eTkFwb"
            content="601"
            fontFamily="indv-mkt-icons"
            fontSize="inherit"
          />
          collapse 2
        </span>
        <aside />
      </button>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="abcdef"
      className="rah-static rah-static--height-zero es-drawer__body sc-ifAKCX bUSXbP"
      id="abcdef"
      role="region"
      style={
        Object {
          "height": 0,
          "overflow": "hidden",
        }
      }
    >
      <div>
        second
      </div>
    </div>
  </div>
  <div
    className="es-drawer__panel third sc-bwzfXH bBIqkK"
  >
    <div
      className="es-drawer__heading"
      id="abcdef"
      role="heading"
    >
      <button
        aria-controls="abcdef"
        aria-expanded={false}
        className="sc-htpNat jDwJgO"
        onClick={[Function]}
      >
        <span>
          <i
            aria-hidden={true}
            className="es-icon sc-bxivhb bCiBaT sc-bdVaJa eTkFwb"
            content="601"
            fontFamily="indv-mkt-icons"
            fontSize="inherit"
          />
          collapse 3
        </span>
        <aside />
      </button>
    </div>
    <div
      aria-hidden={true}
      aria-labelledby="abcdef"
      className="rah-static rah-static--height-zero es-drawer__body sc-ifAKCX HqBXT"
      id="abcdef"
      role="region"
      style={
        Object {
          "height": 0,
          "overflow": "hidden",
        }
      }
    >
      <div>
        third
      </div>
    </div>
  </div>
</div>
`;
